[lldb][progress] Mitigate non-specific LLDB progress reports in Swift

chelcassanova opened the pull request:
When LLDB reports progress on importing Swift modules, it was delivering non-specific progress reports with only the title of "Importing Swift modules" and no details. To report progress on activity within Swift, LLDB first creates a progress report then updates that progress report using a callback that LLDB sets and the Swift compiler invokes when performing a full import.

When LLDB triggers Swift to import modules that were already imported before, Swift will not perform a full import. Since the progress report would've already been displayed, but the callback is never invoked, this leads to the non-specific messages that were being displayed when importing Swift modules.

This commit sets a unique pointer to create a progress report from within the callback function instead of creating the report before setting the callback. It also clears the callback on scope exit instead of setting it to a new progress report.

rdar://122050052
https://github.com/apple/llvm-project/issues/8572

chelcassanova committed:
[lldb][progress] Mitigate non-specific LLDB progress reports in Swift

When LLDB reports progress on importing Swift modules, it was delivering
non-specific progress reports with only the title of "Importing Swift
modules" and no details. To report progress on activity within Swift,
LLDB first creates a progress report then updates that progress report
using a callback that LLDB sets and the Swift compiler invokes when
performing a full import.

When LLDB triggers Swift to import modules that were already imported before,
Swift will not perform a full import. Since the progress report would've
already been displayed, but the callback is never invoked, this leads
to the non-specific messages that were being displayed when importing
Swift modules.

This commit sets a unique pointer to create a progress report from
within the callback function instead of creating the report before
setting the callback. It also clears the callback on scope exit instead
of setting it to a new progress report.

chelcassanova commented:
@swift-ci please test

adrian-prantl reviewed (approved):
Do we not have any tests for this that need to be updated?

chelcassanova committed:
Update test for Swift progress reports

Updates the test for Swift progress reporting by checking that the
"Importing Swift modules" message only appears twice on its own as the
rest of the reports for this category will have details attached.

chelcassanova commented:
Just updated this to update the test.

chelcassanova commented:
@swift-ci please test

chelcassanova commented:
@swift-ci please test windows

chelcassanova commented:
@swift-ci please test

chelcassanova commented:
@swift-ci please test windows

medismailben reviewed (approved):
LGTM!

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3781,20 +3777,22 @@ swift::ModuleDecl *SwiftASTContext::GetModule(const SourceModule &module,
   // Report progress on module importing by using a callback function in
   // swift::ASTContext.
-  Progress progress("Importing Swift modules");
+  std::unique_ptr<Progress> progress;
   ast->SetPreModuleImportCallback(
       [&progress](llvm::StringRef module_name,
                   swift::ASTContext::ModuleImportKind kind) {
+        if (!progress)
+          progress = std::make_unique<Progress>("Importing Swift modules");
```
Are these callbacks always called on the same thread?

chelcassanova replied:
I'm actually unsure of this but I can check.

chelcassanova replied:
I asked Adrian and he noted that while it's not guaranteed that the callbacks will be performed on the same thread, it is guaranteed that they will be called synchronously.
